ospfd/ospf_lsa.c

```diff
@@ -469,6 +469,12 @@ char link_info_set(struct stream **s, struct in_addr id, struct in_addr data,
 }
 
 /* Describe Point-to-Point link (Section 12.4.1.1). */
+
+/* Note: If the interface is configured as point-to-point dmvpn then the other
+ * end of link is dmvpn hub with point-to-multipoint ospf network type. The
+ * hub then expects this router to populate the stub network and also Link Data
+ * Field set to IP Address and not MIB-II ifIndex
+ */
 static int lsa_link_ptop_set(struct stream **s, struct ospf_interface *oi)
```

Comment:
even if I do a shortcut, do we still consider spoke as point to point ?

Reply:
Yes. There is normally no map multicast done for the shortcuts. Spoke was OSPF only communicating with the hub in my testing.

Now I think in principle you could setup the spokes with map multicast dynamic, and then have associations with multiple other routers. Although I did not test this out, and I am not sure how much sense this would make. The spokes learn to do shortcuts through traffic indications from the base.

Reply:
> even if I do a shortcut, do we still consider spoke as point to point ?

Yes, spoke shall still be point-to-point. Note that "point-to-point" is at the OSPF layer which means spoke forms OSPF neighbor adjacency ONLY with hub. Thus, even if spoke-spoke shortcut is established, they won't become OSPF neighbors (as OSPF multicasts are only getting exchanged between hub and spoke)

However, there was one comment in cumulus network doc which said

"OSPF unnumbered is supported with point-to-point interfaces **only** and does not support network statements."

https://docs.cumulusnetworks.com/cumulus-linux-43/Layer-3/OSPF/Open-Shortest-Path-First-v2-OSPFv2/##ospfv2-unnumbered

But, if FRR OSPF point-to-multipoint network type supports unnumbered interfaces without any restrictions then I don't see a reason why we cannot make "ip ospf network point-to-multipoint" work on spoke

Reply:
I did not find a literature saying it is forbidden to do adjacencies with shortcuts ( neither the opposite:)).
I am ok for keeping the code as it is today.

However, to better understand what can be possible, if we agree that it is possible to configure ptomultipoint, to establish adjacencies with shortcuts, then I think there should be a change in the function `nhrp_multicast_foreach()', where each dynamic cache entry would be sent multicast traffic ?

And if we use a dynamic multicast list based on the list of cache entries, which dynamic entries should we choose ?
As a note on [https://github.com/FRRouting/frr/pull/7747/commits/ac95bcefe4492e620c33ab17b2c92c203abc2a5b], I wonder if it is really feasible.

note that I encountered 

```
show ip nhrp cache 
Iface    Type     Protocol                 NBMA                     Flags  Identity
gre1     local    11.255.255.2             -                               -
gre1     nhs      11.255.255.1             10.125.0.1                T    
gre1     dynamic  11.150.0.1               10.125.0.4                TA   
gre1     dynamic  11.255.255.4             10.125.0.4                TA   
```

Reply:
I'm not sure, I followed your suggestion regarding change in 'nhrp_multicast_foreach()'. If you see 'nhrp_multicast_forward()' - this function sends multicast to each dynamic cache entry if "ip nhrp map multicast dynamic" is configured. Do you see any issues there?

Reply:
> gre1     dynamic  11.150.0.1               10.125.0.4                TA   
> gre1     dynamic  11.255.255.4             10.125.0.4                TA

How did you land into this?

I have only seen this happening (i.e. different protocol addresses but same NBMA) when both the devices are behind the same NAT Router. This use case is not supported (Cisco also does not supports this) - i.e. we cannot have multiple spokes behind a same NAT router

Reply:
> > gre1     dynamic  11.150.0.1               10.125.0.4                TA
> > gre1     dynamic  11.255.255.4             10.125.0.4                TA
> 
> How did you land into this?
> 
> I have only seen this happening (i.e. different protocol addresses but same NBMA) when both the devices are behind 

Possibly, this is related to the branch used for testing.
I will retest ac95bce with master branch the nhrp hub use case. and will let you know.

Now I have the confirmation that it should not be allowed.
Thanks

Reply:
the issue is related to master branch.
there are high chances that recent modification  brought on master broken the master branch:
```
north-vm# show ip nhrp shortcut                                                                   |
Type     Prefix                   Via                      Identity                               |
dynamic  11.150.0.0/24            11.255.255.4                                                    |
north-vm# show ip nhrp cache                                                                      |
Iface    Type     Protocol                 NBMA                     Flags  Identity               |
gre1     local    11.255.255.2             -                               -                      |
gre1     nhs      11.255.255.1             10.125.0.1               UT                            |
gre1     dynamic  11.150.0.1               10.125.0.4               UT                            |
gre1     dynamic  11.255.255.4             10.125.0.4               UT                            |
north-vm#  
```

Reply:
@pguibert6WIND Aah, now I understand what you are doing. I did not realized you are trying shortcut. So what you see above is correct output on spoke
11.255.255.4 is the protocol address and 11.150.0.0/24 is subnet you are creating shortcut for.

You should also see "sh ip route nhrp" and you will see three route entries, something like below:

N>* 11.255.255.4/32 [10/0] is directly connected, gre1, weight 1, 00:00:17
N>* 11.150.0.0/24 [10/0] via 11.255.255.4, gre1 onlink, weight 1, 00:00:17
N>* 11.150.0.1/32 [10/0] is directly connected, gre1, weight 1, 00:00:17

You can also verify "ip neigh show" command on linux and you will find neighbor entries for shortcut are correct

**So basically, I do not see anything wrong in above output you posted. That is correct on spoke.**

Reply:
One configuration we've not tested is "ip nhrp map multicast dynamic" on spokes. What we have tested is below:

Hub:  ip nhrp map multicast dynamic
spokes: ip nhrp map multicast X.X.X.X

If we've to support "ip nhrp map mutlicast dynamic" on spokes then we will need to add a flag in 'struct nhrp_cache' indicating that the dynamic entry is for "protocol address" and not for "external subnet". By 'protocol address', I mean spoke's tunnel source interface IP address (11.255.255.4 in your case I assume)

Then, 'nhrp_multicast_forward()' can forward multicast to only those dynamic cache entries which have this 'protocol address' flag set.

Above can be done I believe but this can be taken as enhancement as I don't think many dmvpn installations will use "ip nhrp map multicast dynamic" on spokes

Reply:
I agree @amollad. thanks for feedback. subject closed.